lib_static/open_project/authentication/strategies/warden/doorkeeper_oauth.rb

```diff
@@ -4,65 +4,60 @@ module OpenProject
   module Authentication
     module Strategies
       module Warden
-        ##
-        # Allows testing authentication via doorkeeper OAuth2 token
-        #
         class DoorkeeperOAuth < ::Warden::Strategies::Base
+          include FailWithHeader
+
           def valid?
-            ::Doorkeeper::OAuth::Token
-              .from_request(decorated_request, *Doorkeeper.configuration.access_token_methods)
-              .present?
+            access_token = ::Doorkeeper::OAuth::Token
+                             .from_request(decorated_request, *Doorkeeper.configuration.access_token_methods)
+            access_token.present? &&
+              begin
+                JWT.decode(access_token, nil, false)
```

Comment:
❓ How do the strategy lookup work? Will all the strategies go through `valid?` Which will be the `winning_strategy`?

My question is: do we need to put some JWT validation into the doorkeeper OAuth, just to return `true` on a failed decoding? Or can we manage to change is so, that if a JWT token is sent, DoorkeeperOAuth would still send back "valid", but JWT OIDC will be the `winning_strategy`?

Intention is to get JWT dependency out of a validation for OAuth bearer tokens. (which have no restrictions of pattern)

Reply:
The order defined in https://github.com/opf/openproject/blob/9428a5df4daeeb7ca1138fb0dd0576cf5a0041cc/config/initializers/warden.rb#L18.
The strategies are tried in this order. If a strategy is `#valid?` then its `#authenticate!` is called.  `#authenticate!` has three options: 
1. call `success!`. It means user has been found and next strategies in the order will not be tried.
2. call `fail!`. It means authentication failed for whatever reason and next strategies in the order will not be tried.
3. call `pass` explicitly or do not call `success!` or `fail!`. User has not been authenticated, but next strategies will be tried.

Speaking about JWT dependency. `doorkeeper-oauth` has possibility to use `jwt`s, but we do not use it in our application for now, because of that I decided to rely on bearer token format to distinguish strategies. I would keep it this way. It is a stable way to decided which strategy to use. At least for now.